### ISSUE TITLE
Double-click to select word then drag doesn't continue to select whole words

### ISSUE THREAD

**Issue by baldurk** [jaredpar/VsVim#2383]
In vim if you double-click the mouse, then hold the second click and drag, it will continue to select whole words (this is a typical behaviour in windows at least, not sure about other OSs). At the moment on VsVim the drag will select character-wise.

I think this is probably related to #2368 since before as I understand it was using the VS mouse handling.

Also additionally - I don't think this is vim behaviour, but previously on VS you'd get the same if you control-click and then drag (word-wise selection). It would be nice to get that too since at the moment it just doesn't drag at all. I can understand though if that's out of scope as that's not original vim behaviour, I just have some muscle memory to overcome :smile:.

**Comment by baldurk** [comment]
Another windows/VS-ism: when doing this kind of selection drag the word that you originally double clicked on is always included in the selection. In vim and now VsVim, if you double click on a word then drag to the left, the word gets removed from the selection. It's unintuitive to me, but when testing it that is the vim behaviour.

Having a setting for these would be really nice if it's not too hard to support either behaviour, but I can understand if that's not something you want to get into.

**Comment by ricksladkey** [comment]
The double-click and drag problem is more of the seemingly never ending list of fringe features that need to be implemented once we handle any left-mouse event. I will take a look at fixing that.

The issue of dragging a selection is more problematic, because:

1) It is not supported by gvim, so we would have to add a non-standard setting to control it
1) There is no base functionality we can leverage off of, so it has to be reimplemented (including mouse cursor change, etc.), which is clearly possible, but definitely not trivial

**Comment by baldurk** [comment]
Yeh from reading that PR it seemed like this might be something where once you've taken control you get all the responsibility, you can't do things piecemeal. Which makes sense, but I can definitely appreciate is really annoying.

If it makes it any easier I'm not talking about actually moving selected text around in the document - not sure if that's what you meant by 'dragging a selection'. I'm only talking about various ways of extending the selected region with a drag after some operation, either double click or control click.

I figured I'd ask about matching VS behaviour for double-click and control-click, since without knowing the code it _might_ not be too hard to include and I think gvim's mouse behaviour in this area is a little less useful just because vim has never been mouse centric. If you decide that it's not worth the hassle that's fair enough though, I can see about hacking something in locally.

**Comment by ricksladkey** [comment]
@baldurk Ah, I misread your control-click and drag observation. I will think about that.

**Comment by ricksladkey** [comment]
@baldurk OK, so we cannot change control-click because vim defines `<C-LeftMouse>` as "go to definition". However these mappings seems to do what you want:

```
:noremap <C-LeftMouse> <LeftMouse><2-LeftMouse>
:noremap <C-LeftDrag> <LeftDrag>
```

so we are back to just the first problem, the title of this issue.

**Comment by ricksladkey** [comment]
@baldurk wrote:

> Another windows/VS-ism: when doing this kind of selection drag the word that you originally double clicked on is always included in the selection. In vim and now VsVim, if you double click on a word then drag to the left, the word gets removed from the selection. It's unintuitive to me, but when testing it that is the vim behaviour.

I think I have intuited vim's logic regarding this behavior. In vim, the original highlighted word is *not* included when dragging backwards when `:set selection=exclusive` is in effect, and it *is* included when dragging backwards when `:set selection=inclusive` is in effect. In other words (no pun intended), it works exactly like normal (non-word) dragging works, except that working on character boundaries, it works on word boundaries.

In other words, dragging by character and dragging by word both strictly obey the exclusive/inclusive concept, just applied to different types of units.
